src/main/java/com/hoverla/bring/context/ApplicationContextImpl.java

```diff
@@ -0,0 +1,115 @@
+
+import com.hoverla.bring.annotation.Autowired;
+import com.hoverla.bring.annotation.Bean;
+import com.hoverla.bring.exception.ApplicationContextInitializationException;
+import com.hoverla.bring.exception.NoSuchBeanException;
+import com.hoverla.bring.exception.NoUniqueBeanException;
+import org.reflections.Reflections;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toMap;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class ApplicationContextImpl implements ApplicationContext {
+    private final Map<String, Object> beans = new ConcurrentHashMap<>();
+
+    public ApplicationContextImpl(String... basePackages) {
+        Reflections beanScanner = new Reflections((Object[]) basePackages);
+        Set<Class<?>> beanClasses = beanScanner.getTypesAnnotatedWith(Bean.class, true);
+
+        if (beanClasses.isEmpty()) {
+            return;
+        }
+
+        try {
+            initBeans(beanClasses);
+            postProcess();
+        } catch (NoSuchMethodException | InvocationTargetException | InstantiationException
+                 | IllegalAccessException e) {
+            throw new ApplicationContextInitializationException(
+                format("ApplicationContext initialization has failed: %s", e.getMessage())
+            );
+        }
+    }
+
+    private void initBeans(Set<Class<?>> beanClasses) throws NoSuchMethodException, InvocationTargetException,
+        InstantiationException, IllegalAccessException {
+        for (Class<?> beanType: beanClasses) {
+            String beanName = resolveBeanName(beanType);
+            Object instance = beanType.getConstructor().newInstance();
+            beans.put(beanName, instance);
+        }
+    }
+
+    @SuppressWarnings("java:S3011")
+    private void postProcess() throws IllegalAccessException {
+        Collection<Object> beanInstances = beans.values();
+
+        for (Object beanInstance: beanInstances) {
+            for (Field field: beanInstance.getClass().getDeclaredFields()) {
+                if (field.isAnnotationPresent(Autowired.class)) {
+                    field.setAccessible(true);
+                    field.set(beanInstance, getBean(field.getType()));
+                }
+            }
+        }
+
+    }
+
+    private String resolveBeanName(Class<?> type) {
+        String beanName = type.getAnnotation(Bean.class).value();
+        return isNotBlank(beanName) ? beanName : resolveBeanName(type.getSimpleName());
+    }
+
+    private String resolveBeanName(String typeName) {
+        return typeName.substring(0, 1).toLowerCase() + typeName.substring(1);
```

Comment:
Could we create a class with a name consisting of  one letter?

Reply:
Yes, I added a few test cases to cover that.